lock-api/src/main/java/com/palantir/lock/client/AsyncTimeLockUnlocker.java

```diff
@@ -102,14 +99,20 @@ private void unlockOutstanding() {
         }
     }
 
-    private void schedulePeriodicKickJob() {
-        // This exists to handle a specific race, where transaction A adds itself to outstandingLockTokens
-        // but an already running task in transaction B has read the tokens and is trying to unlock, and
-        // then no transactions follow - the tokens registered by A will not unlock.
-        // Under high continuous volume of transactions, this job is not important.
-        // Also, it won't affect correctness as it is basically doing an empty-set enqueue.
-        scheduledExecutorService.scheduleAtFixedRate(
-                this::scheduleIfNoTaskRunning, 0, KICK_JOB_INTERVAL.getSeconds(), TimeUnit.SECONDS);
+    private Set<LockToken> getOutstandingLockTokenSnapshot() {
+        // Ensure that we acquire the lock for as short as possible (i.e. only 2 writes)
+        Set<LockToken> toUnlock;
+        Set<LockToken> newSet = Sets.newConcurrentHashSet();
+
+        readWriteLock.writeLock().lock();
+        try {
+            toUnlock = outstandingLockTokens;
+            outstandingLockTokens = newSet;
```

Comment:
Are we back to needing a `volatile` modifier on `outstandingLockTokens` since we're not using explicit `synchronized` blocks anymore? I'll try to dig in to understand this a bit better.

Reply:
Looks like this isn't necessary according to the `Lock` interface: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html. These locks must provide the same guarantees as the built-in locks.

Reply:
Yep, I don't think we need the `volatile` modifier here. I'm pretty confident the lock gives us the required memory visibility semantics.

Gory details (there might be some errors here):
- A `ReentrantReadWriteLock` uses an inner-class implementation of [AbstractQueuedSynchronizer](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/AbstractQueuedSynchronizer.html) for its synchronization.
- Internally, this uses volatiles - the release and acquire methods (both exclusive and shared) call AQS methods that perform volatile reads and writes.
- Read-lock lock/unlock and write-lock lock/unlock methods call the AQS methods you'd expect them to call; crucially they are on the *same* synchronizer.
- Here, the volatile variable we're synchronizing on is inside the AbstractQueuedSynchronizer. When we lock the write lock, we are doing a volatile read of some variable that was written to when the last reader unlocked the read lock. So lock/unlock operations have the same semantics as a volatile read _and_ write.

Happy to talk about this in more detail!